sdk/couchbase-core/src/httpx/json_row_stream.rs

```diff
@@ -0,0 +1,79 @@
+use crate::httpx::error::Result as HttpxResult;
+use crate::httpx::json_row_parser::JsonRowParser;
+use bytes::Bytes;
+use futures::stream::{FusedStream, Stream};
+use futures::StreamExt;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+
+#[derive(Debug)]
+pub(crate) struct JsonRowStream<S>
+where
+    S: Stream,
+{
+    state: State,
+    parser: JsonRowParser,
+    stream: S,
+}
+#[derive(Debug, Eq, PartialEq)]
+enum State {
+    Collecting,
+    Done,
+}
+
+impl<S> JsonRowStream<S>
+where
+    S: Stream<Item = Result<Bytes, reqwest::Error>> + Unpin,
+{
+    pub fn new(stream: S) -> Self {
+        Self {
+            state: State::Collecting,
+            parser: JsonRowParser::new(2),
+            stream,
+        }
+    }
+}
+
+impl<S> FusedStream for JsonRowStream<S>
+where
+    S: Stream<Item = Result<Bytes, reqwest::Error>> + Unpin,
+{
+    fn is_terminated(&self) -> bool {
+        matches!(self.state, State::Done)
+    }
+}
+
+impl<S> Stream for JsonRowStream<S>
+where
+    S: Stream<Item = Result<Bytes, reqwest::Error>> + Unpin,
+{
+    type Item = HttpxResult<Vec<u8>>;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<HttpxResult<Vec<u8>>>> {
+        let this = self.get_mut();
+        loop {
+            match this.state {
+                State::Collecting => match this.parser.next() {
```

Comment:
Nice! I think a few comments explaining each case here would be v. helpful.

Reply:
Done!